src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs

```diff
@@ -69,32 +52,48 @@ public NuGetPackageDownloader(DirectoryPath packageInstallDir, ILogger logger =
 
             if (resource == null)
             {
-                throw new NuGetPackageInstallerException(
+                throw new NuGetPackageDownloaderException(
                     string.Format(LocalizableStrings.FailedToLoadNuGetSource, source.Source));
             }
 
-            string nupkgPath = Path.Combine(_packageInstallDir.Value, packageId.ToString(),
-                packageVersion.ToNormalizedString(),
-                $"{packageId}.{packageVersion.ToNormalizedString()}.nupkg");
+            string nupkgPath = downloadFolder == null ?
+                Path.Combine(_packageInstallDir.Value, packageId.ToString(), resolvedPackageVersion.ToNormalizedString(), $"{packageId}.{resolvedPackageVersion.ToNormalizedString()}.nupkg") :
+                Path.Combine(downloadFolder, $"{packageId}.{resolvedPackageVersion.ToNormalizedString()}.nupkg");
             Directory.CreateDirectory(Path.GetDirectoryName(nupkgPath));
             using FileStream destinationStream = File.Create(nupkgPath);
             bool success = await resource.CopyNupkgToStreamAsync(
                 packageId.ToString(),
-                packageVersion,
+                resolvedPackageVersion,
                 destinationStream,
                 _cacheSettings,
                 _logger,
                 cancellationToken);
 
             if (!success)
             {
-                throw new NuGetPackageInstallerException(
-                    $"Downloading {packageId} version {packageVersion.ToNormalizedString()} failed");
+                throw new NuGetPackageDownloaderException(
```

Comment:
Do the NuGet APIs provide more detail that we potentially hide, e.g. let's say the service returns HTTP 503 which may contain retry indicators, etc. Or there's an authentication issue. Do we risk hiding the root cause of the failure from the user?

Reply:
I don't think so so far. I haven't seen info exposed from http layer